File: conformance/xcheck.mjs
// Cross-language conformance check (TypeScript / CANONICAL side).
// Prints (a) canonicalize + content-id of one identical record, and (b) the
// final TrustState after replaying a conformance vector. Its Python twin
// (xcheck.py) prints the same lines for the v0.1 vector; they MUST be
// byte-identical.
//
// Parameterized (advisory A3): the vector file is an OPTIONAL argv.
//   node xcheck.mjs                          -> conformance/vectors.json (v0.1)
//   node xcheck.mjs conformance/vectors-v0.2.json  -> the pooled v0.2 profile
// With no arg the output is byte-identical to the frozen v0.1 baseline. The
// v0.2 branch builds referencePolicyV02(evidence_weights) and turns each
// check's `ev` descriptor into a hash-covered evidence_ref via evRef(...).
import {
  coldStart, update, defaultPolicy, referencePolicyV02,
  open, act, makeCheckRecord, checkpoint, seal, canonicalize, contentId, evRef,
} from "../reference/ts/src/index.ts";
import { readFileSync } from "node:fs";
import { fileURLToPath } from "node:url";
import { dirname, join, isAbsolute, resolve } from "node:path";

const HERE = dirname(fileURLToPath(import.meta.url));
const arg = process.argv[2];
const vecPath = arg ? (isAbsolute(arg) ? arg : resolve(process.cwd(), arg)) : join(HERE, "vectors.json");
const vec = JSON.parse(readFileSync(vecPath, "utf8"));

// v0.2 vectors declare the pooled profile + an evidence_weights table; v0.1
// leaves both undefined and folds under the reference v0.1 mean.
const isV02 = vec.weighting_profile === "recede/ref-weighting-v0.2";
const policy = isV02 ? referencePolicyV02(vec.evidence_weights) : defaultPolicy();
const ACTOR = vec.scope.actor, TASK = vec.scope.task_type, GTS = vec.ground_truth_source;

// (a) One identical record: the pinned record from the vector.
const rec = vec.expected_record_hash.record;
const { id: _i, sig: _s, ...pre } = rec;
console.log("RECORD_CANONICAL " + canonicalize(pre));
console.log("RECORD_ID        " + contentId(rec));

// (b) Replay the vector.
function build(e) {
  const intent = open({ actor: ACTOR, task_type: TASK, proposed_action: "issue refund", declared_risk: e.risk, ts: e.ts });
  const action = act({ intent, operations: ["refund"], result: { ok: true }, ts: e.ts });
  const checks = (e.checks || []).map((c) => makeCheckRecord({
    action, check_kind: c.kind, method: "m", verdict: c.verdict, confidence: c.confidence,
    evidence_refs: c.ev ? [evRef(c.ev.evClass, c.ev.provTier, c.ev.author, c.ev.artifactDigest, c.ev.locator, { mutation: c.ev.mutation })] : [],
    ts: e.ts,
  }));
  const cps = e.decision ? [checkpoint({ warrant_ref: intent.id, actor: ACTOR, reason: "gate", altitude: "full", decision: e.decision, reviewer: "human", ts: e.ts })] : [];
  const outcome = e.result ? seal({ warrant_ref: intent.id, actor: ACTOR, result: e.result, ground_truth_source: GTS, human_touched: cps.length > 0, ts: e.ts }) : undefined;
  return { intent, action, checks, checkpoints: cps, outcome };
}
let st = coldStart(ACTOR, TASK);
for (const e of vec.entries) {
  st = update(st, build(e), policy, { idle_ms: e.idle_ms ?? 0, drift: e.drift ?? 0, now: e.now }).state;
}
console.log("FINAL_TIER       " + st.tier);
console.log("FINAL_SCORE      " + st.score.toFixed(15));
console.log("FINAL_CONFIDENCE " + st.confidence.toFixed(15));
console.log("FINAL_SAMPLES    " + st.sample_count);

File: conformance/xcheck.py
# Cross-language conformance check (Python / MIRROR side).
# Prints (a) canonicalize + content-id of one identical record, and (b) the
# final TrustState after replaying conformance/vectors.json. Its TS twin
# (xcheck.mjs) prints the same lines; they MUST be byte-identical.
import json
import os
import sys

sys.path.insert(0, os.path.join(os.path.dirname(os.path.abspath(__file__)), "..", "reference", "py"))

from recede import Policy, TrustState, Warrant, core, ops  # noqa: E402
from recede.canonical import canonical_serialize, content_id  # noqa: E402

HERE = os.path.dirname(os.path.abspath(__file__))
# Parameterized (advisory A3): the vector file is an OPTIONAL argv. With no arg
# it defaults to conformance/vectors.json (the v0.1 baseline) and the output is
# byte-identical to its TS twin.
_arg = sys.argv[1] if len(sys.argv) > 1 else None
_vec_path = _arg if (_arg and os.path.isabs(_arg)) else os.path.join(os.getcwd(), _arg) if _arg else os.path.join(HERE, "vectors.json")
with open(_vec_path, encoding="utf-8") as fh:
    VEC = json.load(fh)

# Profile-skip guard: the Python mirror only supports the v0.1 reference profile.
# The pooled v0.2 profile is a declared fast-follow (FF1) — print an explicit
# UNSUPPORTED line and skip its replay rather than folding it under v0.1 to a
# silently-wrong state.
_profile = VEC.get("weighting_profile", "recede/ref-weighting-v0.1")
if _profile != "recede/ref-weighting-v0.1":
    print("PROFILE " + _profile + " UNSUPPORTED (py fast-follow)")
    sys.exit(0)

ACTOR = VEC["scope"]["actor"]
TASK = VEC["scope"]["task_type"]
GTS = VEC["ground_truth_source"]

# (a) One identical record: the pinned OUTCOME from the vector.
rec = VEC["expected_record_hash"]["record"]
body = {k: v for k, v in rec.items() if k not in ("id", "sig")}
print("RECORD_CANONICAL " + canonical_serialize(body))
print("RECORD_ID        " + content_id(body))


# (b) Replay the shared vector.
def build(e):
    intent = ops.open(ACTOR, TASK, "issue refund", e["risk"], ts=e["ts"])
    action = ops.act(intent, ["refund"], result="ok", ts=e["ts"])
    checks = tuple(ops.check(action, c["kind"], "m", c["verdict"], c["confidence"], ts=e["ts"]) for c in e.get("checks", []))
    cp = ops.checkpoint(intent.id, "gate", [], "full", decision=e["decision"], reviewer="human", ts=e["ts"]) if e.get("decision") else None
    outcome = ops.seal(intent.id, e["result"], GTS, human_touched=cp is not None, actor=ACTOR, ts=e["ts"]) if e.get("result") else None
    return Warrant(intent=intent, action=action, checks=checks, outcome=outcome, checkpoint=cp)


policy = Policy()
st = TrustState.cold_start(ACTOR, TASK)
for e in VEC["entries"]:
    st = core.update(st, build(e), policy, idle_ms=e.get("idle_ms", 0.0), drift=e.get("drift", 0.0), now=e.get("now"))
print("FINAL_TIER       " + st.tier)
print("FINAL_SCORE      " + format(st.score, ".15f"))
print("FINAL_CONFIDENCE " + format(st.confidence, ".15f"))
print("FINAL_SAMPLES    " + str(st.sample_count))
